docs/source/miscellaneous/doc/adr/0017-transactions3.md

```diff
@@ -0,0 +1,302 @@
+bug. We hope (though don't have strong evidence owing to small sample sizes and extreme difficulty in root causing) 
+that this will reduce the incidence of AtlasDB corruption tickets in general.
+
+Backup and restore workflows, or any use cases that manually manipulate one of the transaction tables will need to be
+aware that new serialized forms exist. In particular, the transactions2 table should not be read in isolation without
+caution - while values are unambiguous in indicating whether they are written with transactions2 or transactions3,
+a user needs to be aware that a STAGING value is not safe to use as though it is final.
+
+It is possible, even likely, that the performance of transactions3 is worse than that of transactions2 (we do strictly
+more work, and values that need to be passed to and from the database are strictly larger). However, we don't expect
+this to add more than a constant amount of overhead, and this should be very small in the steady state. This isn't a
+very fair comparison, in any case, as the transactions2 approach is not correct.
+
+## Appendices
+
+### Proof of Read Consistency
+We first prove a lemma, Lemma 1: the first time the quorum of nodes with the latest writes contains a combination of
+(V, STAGING) and (V, COMMITTED), it is no longer possible to write anything other than (V, STAGING) and (V, COMMITTED).
+
+It is clear that, once the condition is satisfied, a PUE or CAS for value other than V cannot succeed since every read
+will include V as the most recent write. It remains to show that the two puts in the protocols can only be
+PUT(V, COMMITTED).
+
+1. Assume the write protocol can PUT(W, COMMITTED) such that W != V. Then, there must have been a successful
+   PUE(W, STAGING) before it. Since that operation would satisfy the condition of the lemma, implying W=V, the PUE must
+   have occurred after the quorum of nodes with the latest writes already contained V. This is a contradiction since any
+   PUE would have then failed.
+2. Assume the read protocol can CAS((W, STAGING), (W, STAGING)), for W != V. Then, there must have been a (W, STAGING)
+   written after the condition of the lemma was satisfied. Choose the earliest such write, i.e. one that was not
+   replicated from a read. As in 1) this could not have been the result of a PUE, but it also could not have been the
+   result of a CAS because this is the earliest such write and the read part of a CAS would read V as the latest value
+   in any quorum of nodes.
+3. Assume the read protocol can PUT(W, COMMITTED), for W != V. Then, similarly to 1), a successful
+   CAS((W, STAGING), (W, STAGING)) must have occurred beforehand, but after the condition of the lemma was satisfied.
+   This now follows the proof of 2).
+
+We then prove a second lemma, Lemma 2: after a (V, COMMITTED) is written, it is no longer possible to write anything 
+other than (V, STAGING) and (V, COMMITTED).
```

Comment:
under what circumstances (if any - network partition?) would we write `(V, STAGING)` _after_ `(V, COMMITTED)`?

Reply:
Yep, it's a bit tricky. Suppose the cluster is [A, B, C] it's possible to have something like this:

1. There is a network partition and C is isolated.
2. A performs the write protocol: it does PUE of (V, STAGING) on A and B.
3. A performs a PUT of (V, COMMITTED) on A and B. This succeeds on A, but not on B.
4. There is a new network partition and A is isolated, but B and C can now talk to each other.
5. B performs the read protocol; this reads (V, STAGING) from B and empty from C. Because of blocking read repairs, this writes (V, STAGING) to C.

Reply:
nice - thanks!